Magnum Prandium/Assets/ShakyShakeEarthquakeObjectController.cs

```diff
@@ -0,0 +1,82 @@
+    {
+        if (shaking == true && thisObjectsLastShake !=null)
+        {
+            StopCoroutine(thisObjectsLastShake);
+            transform.localPosition = originalPosition;
+            transform.localRotation = originalRotation;
+            shaking = false;
+        }
+        originalPosition = transform.localPosition;
+        originalRotation = transform.localRotation;
+        thisObjectsLastShake = StartCoroutine(ShakeObject2D(duration, magnitudePos, magnitudeRot, fadeIn, fadeOut));
+    }
+
+    IEnumerator ShakeObject2D(float duration, float magnitudePos, float magnitudeRot, float fadeIn, float fadeOut)
+    {
+        shaking = true;
+        Vector3 originalPosition = transform.localPosition;
+        Quaternion originalRotation = transform.localRotation;
+        float currentDuration = 0;
+        float fadeInCompletedTime = duration * fadeIn;
+        float fadeOutStartTime = duration - (duration * fadeOut);
+
+        float currentMagnitudePos=0;
+        float currentMagnitudeRot=0;
+
+        Debug.Assert(fadeIn >= 0 && fadeIn <= 1, "Error, shake fade in is out of bounds(0-1) in: " + this);
+        Debug.Assert(fadeOut >= 0 && fadeOut <= 1, "Error, shake fade out is out of bounds(0-1) in: " + this);
+        Debug.Assert((fadeIn + fadeOut) <= 1, "Error, fadeIn(" + fadeIn + ") + fadeOut(" + fadeOut + ") should be <= 1.0 in: " + this);
+        
+        while (currentDuration < duration)
+        {
+            currentDuration = Mathf.Min(duration, currentDuration + Time.deltaTime);
+            if(currentDuration<=fadeInCompletedTime && fadeInCompletedTime != 0)
+            {
+                currentMagnitudePos = magnitudePos * (currentDuration / fadeInCompletedTime);
+                currentMagnitudeRot = magnitudeRot * (currentDuration / fadeInCompletedTime);
+            }
+            else if(currentDuration>=fadeOutStartTime && fadeOutStartTime != duration)
+            {
+                currentMagnitudePos = magnitudePos * (1 - ((currentDuration - fadeOutStartTime) / (duration - fadeOutStartTime)));
+                currentMagnitudeRot = magnitudeRot * (1 - ((currentDuration - fadeOutStartTime) / (duration - fadeOutStartTime)));
+            }
+            else
+            {
+                currentMagnitudePos = magnitudePos;
+                currentMagnitudeRot = magnitudeRot;
+            }
+            transform.localPosition = new Vector2(originalPosition.x + (Random.Range(-currentMagnitudePos, currentMagnitudePos)), originalPosition.y + (Random.Range(-currentMagnitudePos, currentMagnitudePos)));
+            transform.localRotation= Quaternion.Euler(originalRotation.eulerAngles.x, originalRotation.y, originalRotation.z + (Random.Range(-currentMagnitudeRot, currentMagnitudeRot)));
+            
+            yield return null;
+        }
+
+        shaking = false;
+        transform.localPosition = originalPosition;
+        transform.localRotation = originalRotation;
+    }
+}
```

Comment:
I'll have to run this code to see if it feels good, but Cinemachine has a really good Camera Shake system you are not using. Try adding an impulse listener module to the vcam and triggering an impulse source.

Reply:
The Cinemachine tracker/dolly keeps derping out spitting out weird positional transposed values, even setting damping to 0 the camera just jumps to keep up with the player occasionally, sometimes triggered when doing a point turn right after a dash, but not always.  Setting it to hard locked tracking fixes that for some reason, but then necessitates a dolly object.  If we want I can implement a slurp and retargeting solution pretty easily to re-implement cinemachines soft follow feature and the hard borders?  The impulse module only really adds the ability to trigger shake when it gets hit by stuff, something that mine can also do by making a call in a colision script, and add axis specific frequency generation(which if we feel is important I guess I could actually pretty easily add to mine).  It seems to use the same "random" wave every time for a given frequency though, which is quite annoying.  It is nice that it can automatically do distance calculations, but doing vector2d.distance(vec1,vec2) and feeding that into mine as magnitude when it is called isn't terribly intensive(and if we want I could VERY easily write a version that just takes a transform parameter and it's own and gets a distance to automatically scale magnitude with).  I just don't really see why one is terribly better over the other.  One thing that's nice about mine is it can be put on any object, not just cameras, so if we wanted to make a barrel shake or something we could do that, and my version does actually also support setting attack, sustain, and decay values(I just called mine like duration, fadeIn, and fadeOut).  One thing that is nice about the cinemachine version though is if a second shake is triggered before the first ends it seems to behave by shaking additively.  I could add this with a some work on the script if we feel it's essential?  I can also set up events and listeners if we want to make shock waves propogate from points, and in addition to that, I could even add a delay from the focal point so that things don't start shaking "until the wave reaches them" but just scaling a delay factor of the calculated distance value.  If we go the listener and transmitter route, then we no longer need to even pass in a transform to get the distance as the transmitter script can do that internally(actually that's probably how cinemachine's works if I had to guess).  Idk, If the listener worked on everything, it could be cool, but as far as it only working on cameras I just don't see a whole lot of reason to use most of the extra features.  Now if we could shake in world objects relative to a shock waves origin too then this starts to make a lot more sense for an implementation.... I kind of want to code this now.  I might re-write this over the weekend to do just that.

Reply:
Actually, maybe I will just add camera fade control and inter camera dolly slurps and outright replace cinemachine altogether....

Reply:
You're sacrificing a lot by giving up Cinemachine. First, the smoothing algorithms have special weighted damping functions in place that mimic real-world heavy cameras and tracking systems. They are also designed to eliminate movements that could cause motion sickness. Second, the system is well documented, whereas your system is not, and you are spending a bunch of time replacing a perfectly good system just because you can't solve a strange behavior. And third, you are sacrificing all of your cutscene transition capabilities. That stuff is magic and is especially powerful for this style of game. I'm going to try and reproduce your issues.